src/requestLooper.ts

```diff
@@ -0,0 +1,119 @@
+import { DataFrame, DataQuery, DataQueryRequest, DataQueryResponse, LoadingState } from '@grafana/data';
+import { Observable, Subscription } from 'rxjs';
+
+export interface RequestLoopOptions<TQuery extends DataQuery = DataQuery> {
+  /**
+   * If the response needs an additional request to execute, return it here
+   */
+  getNextQuery: (rsp: DataQueryResponse) => TQuery | undefined;
+
+  /**
+   * The datasource execute method
+   */
+  query: (req: DataQueryRequest<TQuery>) => Observable<DataQueryResponse>;
+
+  /**
+   * Process the results
+   */
+  process: (data: DataFrame[]) => DataFrame[];
+
+  /**
+   * Check if the query should be cancelled
+   */
+  shouldCancel: () => boolean;
+
+  /**
+   * Callback that gets executed when unsubscribed
+   */
+  onCancel: () => void;
+}
+
+/**
+ * Continue executing requests as long as `getNextQuery` returns a query
+ */
+export function getRequestLooper<T extends DataQuery = DataQuery>(
+  req: DataQueryRequest<T>,
+  options: RequestLoopOptions<T>
+): Observable<DataQueryResponse> {
+  return new Observable<DataQueryResponse>((subscriber) => {
+    let nextQuery: T | undefined = undefined;
+    let subscription: Subscription | undefined = undefined;
+    let loadingState: LoadingState | undefined = LoadingState.Loading;
+    let nextRequestDelay = 1; // Seconds until the next request
+    let count = 1;
+    let shouldCancel = false;
```

Comment:
Note, that these are closed over by the observer defined below. E.g. the `shouldCancel` is ultimately how the query editor can communicate to this observable to cancel a running query. See the cancel function on line 191 and the `shouldCancel` function on line 168 in `datasource.ts`.